app/models/project.rb

```diff
@@ -181,6 +181,13 @@ def last_deploy_by_stage
     Deploy.find(found.map(&:id)).select(&:stage).sort_by { |d| d.stage.order }.presence
   end
 
+  def deployed_reference_to_non_production_stage?(reference)
+    stages.joins(deploys: :job).where(
+      deploys: {reference: reference},
+      jobs: {status: 'succeeded'}
+    ).distinct.reject(&:production?).any?
```

Comment:
I'd bet this is pretty expensive to join stage/deploy/job ... maybe get the stage_ids and then do `deploys.where(stage_ids: ids)`

Reply:
also not sure why distinct is needed ... every deploy only has a single stage

Reply:
doing the reject first means it needs to do all the expensive "is this production" checks ... prefer a single any so it can short-circuit

Reply:
if it does not query for commit than `master` will always be ready

Reply:
Three different versions of this query on classic's data which has a ton of deploys: 
```Ruby
    # V3
    non_production_stages = stages.reject(&:production?).map(&:id)
    Deploy.where(stage_id: non_production_stages).joins(:job).where(jobs: {commit: commit, status: 'succeeded'}).any?

    # V2
    jobs_with_commit = Job.where(commit: commit, status: 'succeeded').pluck(:id)
    stages_with_commit_depoyed = Deploy.where(job_id: jobs_with_commit).pluck(:stage_id).uniq
    Stage.where(id: stages_with_commit_depoyed).any? { |stage| !stage.production? }

    # V1
    stages.joins(deploys: :job).where(
      jobs: {status: 'succeeded', commit: commit}
    ).distinct.any? { |stage| !stage.production? }
```

Results: 
```
   user     system      total        real        # queries
V1 .023371   0.013249   0.036620 (  0.119199)        21
V2 0.002848   0.000849   0.003697 (  1.752046)        3
V3 0.002021   0.000603   0.002624 (  0.131441)        1

   user     system      total        real        # queries
V1 0.027942   0.009176   0.037118 (  0.094476)       21
V2 0.013067   0.005522   0.018589 (  1.444021)       12
V3 0.017112   0.000404   0.017516 (  0.142118)       10
```

Looks like V3 is the winner here